src/_pytest/tmpdir.py

```diff
@@ -273,11 +279,16 @@ def tmp_path(
     # Remove the tmpdir if the policy is "failed" and the test passed.
     tmp_path_factory: TempPathFactory = request.session.config._tmp_path_factory  # type: ignore
     policy = tmp_path_factory._retention_policy
-    if policy == "failed" and request.node._tmp_path_result_call.passed:
+    result_dict = request.node.stash[tmppath_result_key]
+
+    # "call" might be skipped so check if it exists first
+    if "call" not in result_dict or (policy == "failed" and result_dict["call"]):
```

Comment:
This now always deletes the tmpdir if there was no `"call"`, no matter what the policy is. That seems odd to me. If I set the policy to `all`, I'd expect those to be kept, even if the test wasn't executed, no?

What about something like:

```suggestion
    if policy == "failed" and result_dict.get("call", True):
```

instead? This would delete them if the test wasn't run, but only if the policy is set to failed.